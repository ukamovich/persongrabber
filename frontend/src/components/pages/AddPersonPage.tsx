

export function AddPersonPage() {
    
      
        return (
      
          <form className="requires-validation" noValidate >
            <h3>Add a new person <span className="label label-default"></span></h3>
              <div className="col-md-12">
                <input type="text" className="form-control mt-3" required placeholder="First Name" name="name" />
                <div className="invalid-feedback">Name field is empty</div>
              </div>
              <div className="col-md-12">
                <input type="text" className="form-control mt-3" required placeholder="Last Name" name="name" />
                <div className="invalid-feedback">Name field is empty</div>
              </div>
              <div className="col-md-12">
                <input type="text" className="form-control mt-3" required placeholder="Email adress" name="email" />
                <div className="invalid-feedback">Email field is empty</div>
              </div>
              <div className="col-md-12">
                <input type="date" className="form-control mt-3" placeholder="Date of birth" name="age" />
                <div className="invalid-feedback">Date field is empty</div>
              </div>
              <div className="form-group">
                <select className="form-select mt-3" name="name" >
                  <option selected disabled value="">Gender</option>
                  <option value="male">Male</option>
                  <option value="female">Female</option>
                </select>
              </div>
              <div className="col-md-12">
                <textarea className="form-control mt-3" placeholder="Biography" name="bio"></textarea>
              </div>
              
            <div className="">
              <button type="submit" id="submit" className="btn btn-primary mt-3 mx-3" >Add Person</button>
              <button type="button" className="btn btn-primary mt-3">Switch</button>
            </div>
          </form>
        );
}
    
    
  
