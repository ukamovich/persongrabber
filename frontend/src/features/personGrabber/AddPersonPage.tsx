import React, { useState, useEffect } from "react";
import fetchGrabber from "../../_helpers/fetchGrabber";

const port = 3001 || process.env.REACT_APP_BACKEND_PORT
const backendURL = `http://localhost:${port}/graphql`

const initialPerson = {
  first_name: "",
  last_name: "",
  email: "",
  birthdate: "",
  gender: "",
  bio: ""
}

interface PersonInterface {
  first_name: string
  last_name: string
  email: string
  birthdate: string
  gender: string
  bio: string
}

function validateForm(person: PersonInterface): boolean {
  for (let [key, value] of Object.entries(person)) {
    if (!value) {
      alert(`${key} cannot be empty!`)
      return false
    }
  }
  return true
}


function AddPersonPage() {

  const [person, setPerson] = useState<PersonInterface>(initialPerson)
  const [id, setId] = useState("")

  const [genders, setGenders] = useState(["Male"])

  useEffect(() => {
    let query = {
      query: `
          query {
              generalPeopleInfo(distinct: "gender") {
                  distinct
              }
          }
      `
  }
  fetchGrabber(query, backendURL).then((res)=>{
      setGenders(res.data.generalPeopleInfo.distinct)
  })
  }, [])

  const handleChange = (event: React.ChangeEvent<HTMLInputElement|HTMLSelectElement|HTMLTextAreaElement>) => {
    setPerson({ ...person, [event.target.name]: event.target.value })
  }


  const submitForm = (event: React.FormEvent<HTMLFormElement>) => {
    // Preventing the page from reloading
    event.preventDefault();

    if (validateForm(person)) {
      let queryBody = {
        query: `
          mutation {
            createPerson(data: {
              first_name: "${person.first_name}"
              last_name: "${person.last_name}"
              email: "${person.email}"
              birthdate: "${person.birthdate}"
              gender: "${person.gender}"
              bio: "${person.bio}"
            }) {
              _id
              first_name
            }
          }
        `
      }
      fetchGrabber(queryBody, backendURL).then(res => {
        alert("Success! New persons id is: " + res.data.createPerson._id)
        setPerson(initialPerson)
        setId(res.data.createPerson._id)
      })
    } else {
      console.log("Err!")
    }
  }


  return (

    <form onSubmit={submitForm} style={{textAlign: "center"}}>
      <h3>Add a new person <span className="label label-default"></span></h3>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" required placeholder="First Name" name="first_name" onChange={handleChange} />
      </div>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" required placeholder="Last Name" name="last_name" onChange={handleChange} />
      </div>
      <div className="col-md-12">
        <input type="text" className="form-control mt-3" required placeholder="Email adress" name="email" onChange={handleChange} />
      </div>
      <div className="col-md-12">
        <input type="date" className="form-control mt-3" placeholder="Date of birth" name="birthdate" onChange={handleChange} />
      </div>
      <div className="form-group">
        <select className="form-select mt-3" name="gender" onChange={handleChange} >
          <option disabled value="">Gender</option>
          {genders && genders.map(el => {return <option value={el} key={el}>{el}</option>})}
        </select>
      </div>
      <div className="col-md-12">
        <textarea className="form-control mt-3" placeholder="Biography" name="bio" onChange={handleChange}></textarea>
      </div>

      <button type="submit" id="submit" className="btn btn-primary mt-3 mx-3" >Submit</button>
      {id && (
        <p>New person id: {id}</p>
      )}
    </form>
  );
}

export default AddPersonPage



