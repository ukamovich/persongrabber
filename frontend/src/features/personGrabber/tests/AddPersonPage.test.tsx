import AddPersonPage from '../AddPersonPage'
import { shallow, mount } from "enzyme"
import React from 'react'

const updateInput = (wrapper: any, instance: any, newValue: any, name:string) => {
    const input = wrapper.find(instance)
    input.simulate('change', {
        target: {value: newValue, name: name}
    })
    return wrapper.find(instance)
}

describe("Testing person creation form", () => {
    test("Renders correctly", () => {
        shallow(<AddPersonPage/>)
    })

    test("Can input values in fields", () => {
        const wrapper = shallow(<AddPersonPage/>)
        const first_name = updateInput(wrapper, '[name="first_name"]', "Jack", "first_name")
        expect(first_name.props().value).toBe("Jack")
        const last_name = updateInput(wrapper, '[name="last_name"]', "Volvo", "last_name")
        expect(last_name.props().value).toBe("Volvo")
        const email = updateInput(wrapper, '[name="email"]', "test@gmail.com", "email")
        expect(email.props().value).toBe("test@gmail.com")
        
    })
})