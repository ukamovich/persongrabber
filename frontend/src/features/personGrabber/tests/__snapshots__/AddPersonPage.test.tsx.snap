// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add person page renders properly 1`] = `
<form
  className="needs-validation"
  noValidate={true}
  onSubmit={[Function]}
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <h3>
    Add a new person 
    <span
      className="label label-default"
    />
  </h3>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      name="first_name"
      onChange={[Function]}
      pattern="[a-zA-ZæøåÆØÅöÖÄäëË]{2,20}"
      placeholder="First Name"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please enter a valid first name
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      name="last_name"
      onChange={[Function]}
      pattern="[a-zA-ZæøåÆØÅöÖÄäëË]{2,20}"
      placeholder="Last Name"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please enter a valid last name
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      name="email"
      onChange={[Function]}
      pattern="^[^\\\\s@]+@[^\\\\s@]+\\\\.[^\\\\s@]+$"
      placeholder="Email adress"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please enter a valid email
  </div>
  <div
    className="col-md-12"
  >
    <input
      className="form-control mt-3"
      max="2021-10-31"
      name="birthdate"
      onChange={[Function]}
      placeholder="Date of birth"
      required={true}
      type="date"
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Please enter a valid date
  </div>
  <div
    className="form-group"
  >
    <select
      className="form-select mt-3"
      name="gender"
      onChange={[Function]}
      value="Male"
    >
      <option
        disabled={true}
        value=""
      >
        Gender
      </option>
      <option
        value="Male"
      >
        Male
      </option>
      <option
        value="Female"
      >
        Female
      </option>
    </select>
  </div>
  <div
    className="invalid-feedback"
  >
    Please select a gender
  </div>
  <div
    className="col-md-12"
  >
    <textarea
      className="form-control mt-3"
      name="bio"
      onChange={[Function]}
      placeholder="Biography"
      required={true}
      value=""
    />
  </div>
  <div
    className="invalid-feedback"
  >
    Bio cant be empty
  </div>
  <button
    className="btn btn-primary mt-3 mx-3"
    id="submit"
    type="submit"
  >
    Add Person
  </button>
  
</form>
`;
